program/src/big_vec.rs

```diff
@@ -0,0 +1,490 @@
+
+        if self.data.len() < end_index {
+            return Err(ProgramError::AccountDataTooSmall);
+        }
+
+        if end_index - start_index != T::LEN {
+            // This only happends if start_index is very close to usize::MAX,
+            // which means that T::LEN should be huge. Solana does not allow such values on-chain
+            return Err(LidoError::AccountListIndexOutOfBounds.into());
+        }
+
+        Ok((start_index, end_index))
+    }
+
+    /// Get element at position
+    pub fn get_mut<T: Pack>(&mut self, index: u32) -> Result<&mut T, ProgramError> {
+        let (start_index, end_index) = self.get_slice_bounds::<T>(index)?;
+        let ptr = self.data[start_index..end_index].as_ptr();
+        Ok(unsafe { &mut *(ptr as *mut T) })
+    }
+
+    /// Removes and returns the element at position index within the vector, shifting all elements after it to the left.
+    pub fn remove<T: Pack + Clone>(&mut self, index: u32) -> Result<T, ProgramError> {
+        let (start_index, end_index) = self.get_slice_bounds::<T>(index)?;
+        let ptr = self.data[start_index..end_index].as_ptr();
+        let value = unsafe { (*(ptr as *const T)).clone() };
+
+        let data_start_index = VEC_SIZE_BYTES;
+        let data_end_index =
+            data_start_index.saturating_add((self.len() as usize).saturating_mul(T::LEN));
+
+        // shift block of memory [end_index..data_end_index] to the left by T::LEN bytes
+        unsafe {
+            sol_memmove(
+                self.data[start_index..data_end_index - T::LEN].as_mut_ptr(),
+                self.data[end_index..data_end_index].as_mut_ptr(),
+                data_end_index - end_index,
+            );
+        }
+
+        let new_len = self.len() - 1;
+        let mut vec_len_ref = &mut self.data[0..VEC_SIZE_BYTES];
+        new_len.serialize(&mut vec_len_ref)?;
+
+        Ok(value)
+    }
+
+    /// Removes an element from the vector and returns it.
+    /// The removed element is replaced by the last element of the vector.
+    /// This does not preserve ordering, but is O(1). If you need to preserve the element order, use remove instead
+    pub fn swap_remove<T: Pack + Clone>(&mut self, index: u32) -> Result<T, ProgramError> {
+        let (start_index, end_index) = self.get_slice_bounds::<T>(index)?;
+        let ptr = self.data[start_index..end_index].as_ptr();
+        let value = unsafe { (*(ptr as *const T)).clone() };
+
+        let data_start_index = VEC_SIZE_BYTES;
+        let data_end_index =
+            data_start_index.saturating_add((self.len() as usize).saturating_mul(T::LEN));
+
+        // if not last element replace it with last
+        if index != self.len() - 1 {
+            unsafe {
+                sol_memmove(
+                    self.data[start_index..end_index].as_mut_ptr(),
+                    self.data[data_end_index - T::LEN..data_end_index].as_mut_ptr(),
+                    T::LEN,
+                );
+            }
```

Comment:
John states [here](https://github.com/solana-labs/solana-program-library/blob/024ba3ad410fef2d31e500c0f1a30db0c222a6a8/stake-pool/program/src/big_vec.rs#L48) that it's not that cheap. If we will ever remove a bunch of values from a vector it will matter. I tried the safe variant, it works well. But I left unsafe just for speed reason. Our workflow is fine for this function and performance will not suffer for now. So I can replace it with safe code.